### PR TITLE
Fix destinationOverrides not being applied to empty aws destination

### PR DESCRIPTION
**Pull request by inFocus7**
# Description

This fixes an issue with `destinationOverrides` not being set if the destination was unset for aws routes.
Adds tests to check the outroute has the applied overrides, and moves tests with "no destinationSpec" to own context.

# Context

While testing more upstream & virtual service routing edge cases I fell upon this.

# Testing

1. Install Gloo with custom `values.yaml` to disable lambda discovery & with fallbackToFirstFunction enabled: `discovery.fdsMode: DISABLED; settings.aws.fallbackToFirstFunction: true`
2. Create an AWS secret 
3. Create an Upstream with a lambda function manually added to it and a `destinationOverride` set (ex. `destinationOverrides.requestTransformation: true`)
4. Create a Virtual Service pointing to upstream, **without** a `destinationSpec set`
5. Port-forward gloo-proxy to 8080
6. `curl -X POST -d '{"foo": "hello world"}' localhost:8080/`

Without this fix you'll see the request go through, but the `destinationOverride` will not take effect.
With this fix, you'll see the request go through as well as the `destinationOverride` taking effect.

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [ ] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [ ] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: projects/gloo/pkg/plugins/aws/plugin.go [hunk 1]
```diff
@@ -175,7 +175,9 @@ func (p *Plugin) ProcessRoute(params plugins.RouteParams, in *v1.Route, out *env
 				// if we fallback to first, we need to have a valid destination
 				contextutils.LoggerFrom(params.Ctx).Debug("no destinationSpec set with fallbackToFirstFunction enabled, processing as aws route")
 				dest = &v1.DestinationSpec{
-					DestinationType: &v1.DestinationSpec_Aws{},
+					DestinationType: &v1.DestinationSpec_Aws{
+						Aws: &aws.DestinationSpec{},
```
Comment (inFocus7) [inline projects/gloo/pkg/plugins/aws/plugin.go:179]: Adding this `Aws`, so that when it's passed to `perRouteConfigGenerator`, it isn't `nil` and the destinationOverrides get applied to it.

File: projects/gloo/pkg/plugins/aws/plugin.go [hunk 2]
```diff
@@ -199,7 +201,15 @@ func (p *Plugin) ProcessRoute(params plugins.RouteParams, in *v1.Route, out *env
 				return nil, err
 			}
 			// should be aws upstream
-			return p.perRouteConfigGenerator(p.settings, awsDestinationSpec.Aws, lambdaSpec)
+			awsLambda, err := p.perRouteConfigGenerator(p.settings, awsDestinationSpec.Aws, lambdaSpec)
+			if err == nil {
+				// set the `destinationSpec` to be `awsDestinationSpec` after being modified in `GenerateAWSLambdaRouteConfig`
+				// doing it this way allows us to keep the original destinationSpec with the updated awsDestinationSpec added onto it
+				dest.DestinationType = awsDestinationSpec
+				spec.DestinationSpec = dest
```
Comment (inFocus7) [inline projects/gloo/pkg/plugins/aws/plugin.go:209]: Setting the `spec`s `DestinationSpec` to the updated values so it still takes effect. 

TODO: think about logic. when `destinationSpec` _was_ set, the destinationOverrides were set correctly without needing to do this. So maybe (some of) this logic isn't necessary?

Comment (inFocus7) [reply projects/gloo/pkg/plugins/aws/plugin.go:207]: Did not need to set `dest.DestinationType = awsDestinationSpec`, as it's modified when passed to `perRouteConfigGenerator`

File: projects/gloo/pkg/plugins/aws/plugin.go [hunk 3]
```diff
@@ -199,7 +201,13 @@ func (p *Plugin) ProcessRoute(params plugins.RouteParams, in *v1.Route, out *env
 				return nil, err
 			}
 			// should be aws upstream
-			return p.perRouteConfigGenerator(p.settings, awsDestinationSpec.Aws, lambdaSpec)
+			awsLambda, err := p.perRouteConfigGenerator(p.settings, awsDestinationSpec.Aws, lambdaSpec)
+			if err == nil {
+				// set the `destinationSpec` to be the original destinationSpec with the updated awsDestinationSpec added onto it
+				spec.DestinationSpec = dest
```
Comment (nfuden) [inline projects/gloo/pkg/plugins/aws/plugin.go:207]: I like that this solves the issue we have been seeing but in general side affects of this manner are a bit scary.
It seems that this is correct given what we talked about above where we know that this is definitely an aws destination (as the upstream is aws) but can we call this out a bit more loudly in the function signature or on the plugin?

Comment (inFocus7) [reply projects/gloo/pkg/plugins/aws/plugin.go:219]: Not sure if you're also content with the updated comment you left. I am, so if you are as well we can resolve this, else I can add some more to it!

**Comment by inFocus7** [comment]
/kick flake #7410 [Details](https://storage.googleapis.com/solo-public-build-logs/logs.html?buildid=928c3782-5478-44bd-b982-2ccbe8e6288f)

File: projects/gloo/pkg/plugins/aws/plugin.go [hunk 4]
```diff
@@ -165,40 +165,60 @@ func (p *Plugin) ProcessUpstream(params plugins.Params, in *v1.Upstream, out *en
...
+			// local variable to avoid side effects for calls that are not to aws upstreams
 			dest := spec.GetDestinationSpec()
-			// check if it's aws destination
-			// it is incorrect to set lambda functionality on routes that do not have a lambda function specified unless we fallback
-			if dest == nil {
-				if p.settings.GetFallbackToFirstFunction().GetValue() == false {
-					return nil, nil
-				}
-				// if we fallback to first, we need to have a valid destination
+			tryingNonExplicitAWSDest := spec.GetDestinationSpec() == nil && p.settings.GetFallbackToFirstFunction().GetValue()
```
Comment (jenshu) [inline projects/gloo/pkg/plugins/aws/plugin.go:170]: when fallback is true, are we validating anywhere that there exists only one lambda function? if there's more than one, should we error?

Comment (inFocus7) [reply projects/gloo/pkg/plugins/aws/plugin.go:170]: I don't believe we check that there's only one lambda fn. The fallback falls back to first lambda function found on upstream, so it doesn't necessarily matter if there's more than one, as long as there is 1 to fall back on.

relevant code
- [ensuring there is at least 1 function](https://github.com/solo-io/gloo/blob/c16e282531adba0ae40558c227dc875719327e3e/projects/gloo/pkg/plugins/aws/plugin.go#L367-L369)
- [using the first function found as fallback](https://github.com/solo-io/gloo/blob/c16e282531adba0ae40558c227dc875719327e3e/projects/gloo/pkg/plugins/aws/plugin.go#L380-L389)

Comment (jenshu) [inline projects/gloo/pkg/plugins/aws/plugin.go:170]:
```suggestion
			tryingNonExplicitAWSDest := dest == nil && p.settings.GetFallbackToFirstFunction().GetValue()
```

**Review by jenshu — DISMISSED** [review]

**Comment by inFocus7** [comment]
/kick flake #6108 [Details](https://storage.googleapis.com/solo-public-build-logs/logs.html?buildid=91dd2f75-fa56-4015-893f-d71ad91a71fa)

**Review by nfuden — APPROVED** [review]
